### ISSUE TITLE
`Type 'string' cannot be used to index type 'T'` when indexing a generic function parameter

### ISSUE THREAD

**Issue by saschanaz** [microsoft/TypeScript#47357]
# Bug Report

### 🔎 Search Terms

Type 'string' cannot be used to index type 'T'.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about index signature

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdEAlbCOAJyLRKhpjAHNEAfREgTwFsARnAA2AGkQBDMDwB8MgBRQJNVtigAuRBnEAHTY2ZtO3fkOEBKRAG8AUIkRKVagNo6AugG47D5aqiu3RABeRAAiUK8AXyA)

### 💻 Code

```ts
function foo<T extends Record<string | symbol, any>>(target: T, p: string | symbol) {
  target[p]; // This is okay
  target[p] = ""; // This errors
}
```

### 🙁 Actual behavior

```
(parameter) target: T extends Record<string | symbol, any>
Type 'string' cannot be used to index type 'T'.(2536)
Type 'symbol' cannot be used to index type 'T'.(2536)
```

### 🙂 Expected behavior

Either:

1. It should work, as the getter already works
2. Or at least the error message should be changed as it's misleading, indexing itself has no problem.

**Comment by MartinJohns** [comment]
The error is correct. You're not extending objects only having `string` typed properties, you're extending objects where the properties extend `string `. `type Test = { a: 'hello' }` is a type that matches your constraint, but assigning an empty string to the property `a` would **not** be correct.

**Comment by saschanaz** [comment]
The error can't be correct, since it *can* be used to index type T. It's at least misleading.

**Comment by MartinJohns** [comment]
We can argue that the error message is misleading and bad. But that an error happens is definitely correct. As my example demonstrates, assigning an empty string would be invalid.

Getting is not an issue here, because you're not using the value for anything. But in your setter you try to assign a potentially incompatible type.

**Comment by saschanaz** [comment]
I'm not sure why `{ a: 'hello' }` is correct and `{ a: '' }` is not 🤔

**Comment by MartinJohns** [comment]
The **type** of the property `a` is `"hello"`, not the **value**. It's a property that can **only** accept the string literal `"hello"`, but it can not accept the string `""`. It's more common used with union types, e.g. `"a" | "b"`. Meaning it can only be the string `"a"`, or the string `"b"`, but not any other string.

```typescript
type Test = { val: "a" | "b" }
const t: Test = { val: "a" }
t.val = "b" // is okay
t.val = "" // is not okay
```

**Comment by saschanaz** [comment]
Oh thanks, now I see. The value type can be anything.

**Comment by saschanaz** [comment]
But:

```ts
function foo<T extends Record<string, any>>(target: T, p: string) {
  target[p];
  target[p] = "hello";
}

function foo2(target: Record<string, any>, p: string) {
  target[p];
  target[p] = "hello";
}

type Test = { val: "a" | "b" }
const t: Test = { val: "a" }

foo(t, "foo");
foo2(t, "foo");
```

Maybe this is just what we should expect by using `any`? 🤔

**Comment by fatcerberus** [comment]
A useful way to think about generics is that you're writing a function that works for *all possible types that `T` could be*; if this isn't true, you get an error in the implementation.  You get no error at the call site, though, because there is no obvious issue with the *call* to `foo()` - `Test` is assignable to `Record<string, any>` and `"foo"` is a `string`.  Obviously `target[p] = "hello"` is not a valid thing to do with a `Test` (or any number of other things you can pass to `foo`), so you get the error inside the function.

Technically the call of `foo2()` *should* be an error, and remains unsound even if you change the `any` to `unknown`, but that's a separate issue: TS treats object types covariantly.  It's unsound for the same reason assigning a `string[]` to a `(string | number)[]` is unsound.  The type system doesn't really model mutation well.

**Comment by fatcerberus** [comment]
If you don't want your random object types to be assignable to `Record<string, T>` you can use interfaces:

```ts
function foo(target: Record<string, unknown>, p: string) {
  target[p];
  target[p] = "hello"; // no error
}

interface Test { val: "a" | "b" }
const t: Test = { val: "a" }

// type error, `Test` has no index signature
foo(t, "foo");
```

Note that using `Record<string, any>` suppresses even this error.

**Comment by Tinadim** [comment]
@fatcerberus Even though I went through the examples and discussion above, I'm not sure I understand exactly what is the root cause of this error. Check the following example:
```typescript
interface Foo {
    [key: string]: number;
}

function test<T extends Foo>() {
    const a = {} as T;
    a['test'] = 1234; // <------------ error I get is: "test can't be used to index type T"
}
```
but why can't `test` be used? it **is** a string, and the constraint of T is `extends Foo`, where the interface for `Foo` specifies that the index signature is a string

**Comment by Tinadim** [comment]
got it, thanks @fatcerberus!

**Comment by Terence625** [comment]
So, what's the good practice to index the generic type T inside the function?

**Comment by houfeng0923** [comment]
So , it need a graceful way to  fix it .

[demo playground](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/FAUwHgDg9gTgLgAjgTwiBAFKAbbBhKAOwDMBLAcwQF4EBvYBBAVxmwC4EBnOGUw8gNwMEfOCBgA3AIbsEhJgFsARuIEJgAXyGNQkWIgAmIYlKbZEAY2xTOnTDmwAeYXgTgxhA3doBtANIihFw8fOQAuhxYuAQkFBrAAHyM9IwWRGTknACEHHjaCGmE3DBMFnCwABSFGZy5AJR0woxwABaknAB01RR2NN2Z+fHCUgYGMRkVIBIghHAcxaEANAXpFJEO4xQNKYxIbZ39nD5TM3Bh1Cuxgk0I2CCIUhet7V2rmcfTs2GDmgxAA)

**Comment by verikono** [comment]
Just a thought for anyone who finds themselves here. I ended up using a one liner taking into account what @MartinJohns  said about "extending".

```
function foo<T extends Record<string | symbol, any>>(target: T, p: string | symbol) {
  target[p]; // This is okay
  target[p] = ""; // This errors
  target = Object.assign({p:""}); // this is okay to.
}
```
be aware target is entirely replaced with a new object.

**Comment by Stevenic** [comment]
@fatcerberus my question is really simple... How can I say that `T` is a plain old JavaScript object that can have any fields keyed by a string? If I can't use `Record<string, any>` to do that what should I use? I'm tired of hitting this error.

**Comment by DiamondMofeng** [comment]
`Reflect.set(obj, key, 1)` works. But I'd also love to find a way to use `obj[key]=1` directly

**Comment by fires3as0n** [comment]
What human intuitively expects when writing `function<T extends Record<string, any>(arg: T)` is `arg` to be of a dictionary type with the ability to additionally specify its exact shape.
From what I understand after reading this thread - unfortunately TypeScript is broken and does not work that way. Generic type can extend a type but the result is unpredictable.

**Comment by rmarscher** [comment]
It would be nice if an `in` operator check or the upcoming `Object.hasOwn` could be used to guard/narrow so that accessing via index by that value then works. Something like this:

``` 
interface A {
	a: string;
}
interface B {
	b: string;
}

export function test<T extends object>(x: T, field = 'a') {
	if (field in x) {
		return x[field];
	}
	// if (Object.hasOwn(x, field)) {
	// 	return x[field];
	// }
	return undefined;
}

const valueA: A = { a: 'yes' };
const valueB: B = { b: 'no' };
test(valueA);
test(valueB);
```

But that code still errors with `No index signature with a parameter of type 'string' was found on type '{}'.`

I don't want to add a `[k: string]: unknown;` definition to my interfaces because then I lose type checking when trying to access an unexpected property.

**Comment by Peeja** [comment]
@Stevenic

> How can I say that `T` is a plain old JavaScript object that can have any fields keyed by a string? If I can't use `Record<string, any>` to do that what should I use? I'm tired of hitting this error.

You *can* use `Record<string, any>` (in fact that's exactly what that means, although `Record<string, unknown>` would be safer). In the example in this issue, the type of `target` isn't `Record<string, any>`, it's some specialization of `Record<string, any>`. It *could* be `Record<string, any>`, but it also could be `Record<"only this", any>`. Clearly _this_ shouldn't be okay:

```ts
function foo(target: Record<"only this", any>, p: string) {
  // We don't know that `p` is an acceptable key -- namely, `"only this"`.
  target[p] = "";
  // You also can't read with `p`, for the same reason.
  target[p]; // You can 
}
```

If you want to treat the argument as any record with string keys, you don't need a generic at all. You can just use that type:

```ts
function foo(target: Record<string, any>, p: string) {
  // Now `p` is just fine as a key.
  target[p] = "";
  target[p];
}
```

Now, there are two things I still don't understand, and which look like bugs to me on the face of them, but I haven't thought about these long:

```ts
function foo1<T extends Record<string, "value in the record">>(target: T, p: string) {
  // `v` is typed as `"value in the record"`, even though `p` may not be a valid key for a particular `T`.
  const v = target[p];
}

function foo2<T extends Record<string, string>>(target: T, p: keyof T) {
  // This assignment isn't allowed, even though `p` is typed as a valid key for any given `T`.
  target[p] = "";
}
```

([playground](https://www.typescriptlang.org/play?ts=4.5.4#code/GYVwdgxgLglg9mABMOcCMAeAKogpgDylzABMBnRAJVwjgCcSMyo6YwBzAGkQCIA3AIYAbELkRtEUABZi6NeiR4A+JQAooAuu1xQAXIizcADvuasOASkQBvAFCJEAekeIABn1fiKUAJ5HcJIgCFK78wqLiSNKy8gw8rty4fMSSUnAg7FJuRp4AtgI+iGBwUIgARmICiIJCMIEA1riFKHRBiEaasBAgQppuWK4AdPaItGDM1YgAvJKa2lAA2kYAugDctgC+tragkLAIyKgATNh4hMTkVLGMZmxciLccKupzOvqG7fqNPnDABlZ2BzOAxSGAUYJkGDsMC5YilMFgADkpWEQjgAHcAolklE0hksq4cl5JH4AkFwdVhHVEN9Dq0BGBCuwYDj+kMRhotDolstprweOstkA))

**Comment by gabritto** [comment]
Here goes a sort of personal summary of the situation, based on my recent investigation of this issue.
Let me first say that index signatures are a bit confusing and inconsistent with the behavior of regular property declarations, so all the questions in this issue are very valid.

Now, why is the error happening?
The error happens because we have a rule that, when you are *writing*/*assigning* to an element access expression, you cannot use an index signature present in the constraint of a generic type. However, you can use that same index signature if you are *reading* from that same element access expression:

```ts
function foo<T extends Record<string, string>>(target: T, p: string) {
  target[p]; // This is okay because we're reading
  target[p] = ""; // This errors because we're writing to `target[p]`
  delete target[p]; // Confusingly, this doesn't error
}
```

If we think about it, writing `target[p] = ""` is *not* safe. Just imagine if we instantiate `T` with `Record<string, "hello">`, which is a more specific type than `Record<string, string>`: the type of `target[p]` would be `"hello"`, but we would be assigning `""` to it.

Ok, so we have a justification for not using the index signature of the constraint `Record<string, string>` of `T` when *writing*, so all is good, right? No, because in similarly unsafe situations, we **do** allow using an index signature or a property declaration of a type parameter's constraint when writing:

```ts
// Indirect assignment to index signature
function foo3<T extends Record<string, { a: string }>>(target: T, p: string) {
  target[p] = { a: "hi" }; // Error
  target[p].a = "hi"; // Same as above, but no error!
}

// Direct assignment with regular property declaration
function foo4<T extends { a: string }>(target: T) {
  target.a = "wow";
}
foo4<{ a: "not wow" }>({ a: "not wow" }); // Bad, will write "wow" to `a` when we expect "not wow".
```

So, from `foo3` above we see that we're not very rigorous about not using an index signature on the left-hand side of an assignment.
And from `foo4` above we see that regular property declarations (i.e. `{ a: string }`) can potentially cause the same type violation/unsafety that we used to justify banning index signatures when writing, but we allow them just fine!


In conclusion, the rules are not very consistent, and I don't have a precise explanation of why it works the way it does, but it probably has to do with how convenient it is to allow the potentially unsafe pattern vs disallowing patterns that are likely to cause bugs.

**Comment by fatcerberus** [comment]
@gabritto I mean, `foo3` is technically consistent under the current rules since `target[p]` is a *read*.  The fact that you then go on to unsoundly mutate the object you pulled out is irrelevant (yes, it's unsound, but that's just object covariance for you; the index signature is out of the equation by this point).

I would *like* to argue that `target[p]` should *always* be an error even when reading, but that would likely be a massive breaking change so I don't expect that proposal to get much traction 😜

**Comment by gabritto** [comment]
> @gabritto I mean, `foo3` is technically consistent under the current rules since `target[p]` is a _read_. The fact that you then go on to unsoundly mutate the object you pulled out is irrelevant (yes, it's unsound, but that's just object covariance for you; the index signature is out of the equation by this point).

Yes, the index signature rule is consistent with our definition of read, but that's a limited definition in the first place. I was trying to make a point that the two statements accomplish a very similar thing, and yet we disallow one and not the other. In the end, even with the index signature rule in place, you might still mutate something whose type was dictated by the index signature. Maybe that's better expressed as "the rule is incomplete". And yes, there are other rules at play as well that cause this to be permitted.

**Comment by 9a8ri3L** [comment]
> But:
> 
> ```ts
> function foo<T extends Record<string, any>>(target: T, p: string) {
>   target[p];
>   target[p] = "hello"; // error
> }
> 
> function foo2(target: Record<string, any>, p: string) {
>   target[p];
>   target[p] = "hello"; // no error
> }
> 
> type Test = { val: "a" | "b" }
> const t: Test = { val: "a" }
> 
> // no error at all
> foo(t, "foo");
> foo2(t, "foo");
> ```
> 
> Maybe this is just what we should expect by using `any`? 🤔

If I correctly understand, I can say the following, (Please anyone can correct me if there's something wrong):

Regarding to the above example, what purpose we have here to add types to the `foo` function?

1. Constrain `p` to the `target` keys. [level 1]
2. Constrain ` target[p] ` value to be ` "a"|"b" `. [level 2]
3. Constraint ` target[p] ` value to be ` "a"|"b" ` based on `p` argument value. [level 3]
4. ... and others

Let's start from [level 3] - because here's the point.

TS has no idea about the literal/union type of `val` in ` interface Test ` which is ` target[p] `.

So for a reason we should do one of the following: 

1. Explicitly tell TS that ` p ` is a key in ` target ` while assigning a value to it in the implementation of ` foo `. 
2. Explicitly tell TS that we assigning `target[p]` to a value belongs to the `target` Record.
3. Extract exactly the `Test` type to the `target` type.

I don't see any reason here for a generic `T` to extend a `Record`, while `Record` itself should receive its arguments from an extended `<K,V>` generics.

```ts
interface Test { val: "a" | "b", bar: 'foo' | 'baz' }
const t: Test = { val: "a", bar: 'baz' }

// level 3
function foo<K extends keyof Test, V extends Test[K]>(target: Record<K, V>, p: K, value: V) {
  target[p]; // okay
  target[p] = value // okay
  return target
}
console.log(foo(t, "val", 'a')) // p constrained to "val"|"bar", and target[p] constrained to the provided p.


// level 2
function _foo<K extends string,V extends string>(target: Record<K,V>, p: K) {
  target[p]; // okay
  target[p] = "" as V; // okay - assertion as V, regardless of type Test { val: "a" | "b", bar: 'foo' | 'baz' }
  return target
}
console.log(_foo<keyof Test, Test[keyof Test]>(t, 'val')) // constrained p "val"|"bar" 


// level 1
function __foo(target: {[k: string]: any}, _p: string) {
  const p = _p as keyof typeof target // cast - assertion p is a key in target
  target[p]; // ok
  target[p] = "abc"; // ok
  return target
}
console.log(__foo(t, 'def')) // non-constrained p

```

[Playground](https://www.typescriptlang.org/play?ts=4.5.4#code/JYOwLgpgTgZghgYwgAgCoQM5mQb2QNzgBsAuZAIjnOQB8KAjcgGmXrijIHIYB7Hz2sk5sAXgIC+AKAQ8QWZGDLp5AXlwFiZSs1bsuoiZMkB6Y8iIR8EIsgDMkmAFcQCMMFnJePADwBpZBAAHpAgACYYyADWEACePDBomGAsAGoBwRBhEcpgANq+ALoAfAAUYOwA5hCKyABKEDJQoX6pRSwADmS+LIREjhBkKQCUuJLICpXVue0FANzIpsg8kXAxYxNQVXkzyGq9-Qtmy6vI61DVjlAgG1uSUjJyPBYAdEQ8FSVeZSzkvTqccE4QxGi3ayAeWCgcFAEFCCh4FD+NHIbCgOjgYRuUx2ELAUJhcLACLAAAsUO0oDx8MBQrDkO1nkYTGYLFYbAAmBzOVzua4AfS8fnSIXCyEhoAqTDSQRFEXFIAqpXKm2qZHqjWa3RSbXpXRGOHWyq20zmhyWKzW4yN2IKuwo1DgERS80WxxiyAAtMhHRhoG4PI7kCkWOcKuxQhYMBF4goYu0UDl1L0tFRBCidKiuF4BHRhHAxMgpONzmBLtdrWA7tJZBgnhBXu8SgK+N5onEEjkWDlcm2Yzlit8hL0gSCzLj8SA6WDfsRyMjUdQmYtWdZkABGLkuf385s8MqTGo4HtkeUVApkDExcQsPmdMV4iX69a4+l22-eiK9hJgOMQGMVs0EEdbAvSCdoiGABBgDAIh3TBYAIjgKJYmQUAsUrK0DxNF0jkiQ0sJ2NRKHoBByBw80zguK50KrCE6wbD4+V3QcARHM0QFkD1x2hSc4XaSQgA)

And that's exactly what I'm (as an old school) waiting from TS, while I'm learning :)
